rambo/app.py

```diff
@@ -391,109 +387,46 @@ def up(ctx=None, provider=None,  guest_os=None, ram_size=None, drive_size=None,
-        utils.warn(msg)
-
-    ## drive_size
-    if drive_size not in iter(SETTINGS['SIZES'].values()):
-        msg = ('DRIVE Size "%s" is not in the DRIVE sizes list.\n'
-               'Did you have a typo? We\'ll try anyway.\n'
-               'Here is as list of avalible DRIVE sizes:\n\n'
-               % drive_size)
-        for supported_drive_size in iter(SETTINGS['SIZES'].values()):
-            msg = msg + '%s\n' % supported_drive_size
-        utils.warn(msg)
-
-    if machine_type:
-        if provider in ('docker', 'lxc', 'virtualbox'):
-            msg = ('You have selected a machine-type, but are not using\n'
-                   'a cloud provider. You selected %s with %s.\n'
-                   % (machine_type, provider))
-            utils.abort(msg)
-        set_env_var('machinetype', machine_type)
+    ## Save Initial command list
+    with open(os.path.join(get_env_var('TMPDIR_PATH'), 'instance.json'), 'w') as fp:
+        json.dump({'params': params}, fp, indent=4, sort_keys=True)
 
+    ## Option Handling - These might modify the params dict or set env vars.
+    params = options.provider_option(params)
+    params = options.guest_os_option(params)
+    params = options.size_option(params) # both ram and drive size
+    params = options.machine_type_option(params)
 
-    ## Provider specific handling.
-    ## Must come after all else, because logic may be done on env vars set above.
-    if provider == 'digitalocean':
-        providers.digitalocean()
-    elif provider == 'docker':
-        providers.docker()
-    elif provider == 'ec2':
-        providers.ec2()
+    ## Save Specs (same `params` now modified)
+    with open(os.path.join(get_env_var('TMPDIR_PATH'), 'instance.json')) as fp:
+        data = json.load(fp)
+    data.update({'specs': params})
+    with open(os.path.join(get_env_var('TMPDIR_PATH'), 'instance.json'), 'w') as fp:
+        json.dump(data, fp, indent=4, sort_keys=True)
 
+    ## Provider specific handling.
+    ## Must come after all else, because logic may be done on params above.
+    if params['provider'] == 'digitalocean':
+        vagrant_providers.digitalocean()
+    elif params['provider'] == 'docker':
+        vagrant_providers.docker()
+    elif params['provider'] == 'ec2':
+        vagrant_providers.ec2()
+    elif params['provider'] == 'gce':
+        gce.up(**params)
 
     ## Add straight pass-through flags. Keep test for True/False explicit as only those values should work
     cmd = 'up'
-    if provision is True:
+    if params['provision'] is True:
         cmd = '{} {}'.format(cmd, '--provision')
-    elif provision is False:
+    elif params['provision'] is False:
         cmd = '{} {}'.format(cmd, '--no-provision')
 
-    if destroy_on_error is True:
+    if params['destroy_on_error'] is True:
         cmd = '{} {}'.format(cmd, '--destroy-on-error')
-    elif destroy_on_error is False:
+    elif params['destroy_on_error'] is False:
```

Comment:
I'm not sure I like how we're getting state from these dictionaries all the time.